### PR TITLE
Port Core/Spells: Cooldown updates (6.x branch)

### PR DESCRIPTION
**Pull request by ariel-**
Core/Spells: Cooldown updates

Updates #14418

### CODE REVIEW

**Comment by ccrs** [comment]
:o NICE!

File: src/server/game/DataStores/DBCEnums.h [hunk 1]
```diff
@@ -362,7 +362,8 @@ enum SkillRaceClassInfoFlags
 enum SpellCategoryFlags
 {
     SPELL_CATEGORY_FLAG_COOLDOWN_SCALES_WITH_WEAPON_SPEED   = 0x01, // unused
-    SPELL_CATEGORY_FLAG_COOLDOWN_STARTS_ON_EVENT            = 0x04
+    SPELL_CATEGORY_FLAG_COOLDOWN_STARTS_ON_EVENT            = 0x04,
+    SPELL_CATEGORY_FLAG_COOLDOWN_EXPIRES_AT_DAILY_RESET     = 0x08
```
Comment (Shauren) [inline src/server/game/DataStores/DBCEnums.h]: thats a new flag in cata - baaaaaaad

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -20835,41 +20636,7 @@ void Player::VehicleSpellInitialize()
...
-        }
-        else
-        {
-            data << uint16(0);
-            data << uint32(cooldown);
-            data << uint32(0);
-        }
-    }
-
+    //vehicle->GetSpellHistory()->WritePacket(&petSpells);
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: commented out? (should write <Pet> block)

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -20923,6 +20690,8 @@ void Player::CharmSpellInitialize()
     }
 
     data << uint8(0);                                       // cooldowns count
+    //Cooldowns
+    //charm->GetSpellHistory()->WritePacket(&petSpells);
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: commented out? (should write Pet cooldown block)

**Comment by ariel-** [comment]
Should it be implemented for vehicle only? as I see CharmSpellInitialize sends a 0 cooldown count.

**Comment by Kittnz** [comment]
What needs to change for this PR to get accepted?

**Comment by MitchesD** [comment]
We can’t automatically merge this pull request.

**Comment by ariel-** [comment]
Bot currently at home, will fix those later and squash; I also noticed that
the "infinity cooldown check" == 15 days is missing and causes some issues
with spells with DISABLED_WHILE_ACTIVE, so I need to put that back in.

2015-04-10 9:46 GMT-03:00 MitchesD <notifications@github.com>:

**Comment by ariel-** [comment]
Done-

File: src/server/game/Entities/Player/Player.cpp [hunk 4]
```diff
@@ -12425,10 +12256,13 @@ Item* Player::EquipItem(uint16 pos, Item* pItem, bool update)
                 {
                     m_weaponChangeTimer = spellProto->StartRecoveryTime;
 
-                    GetGlobalCooldownMgr().AddGlobalCooldown(spellProto, m_weaponChangeTimer);
+                    GetSpellHistory()->AddGlobalCooldown(spellProto, m_weaponChangeTimer);
 
-                    WorldPacket data;
-                    BuildCooldownPacket(data, SPELL_COOLDOWN_FLAG_INCLUDE_GCD, cooldownSpell, 0);
```
Comment (joschiwald) [inline src/server/game/Entities/Player/Player.cpp]: restore the `BuildCooldownPacket` part

**Comment by ccrs** [comment]
again, such a nice work :3

**Comment by ariel-** [comment]
Nope, the sql implemented a wod specific spell charges save system for players and spells (also there are different opcodes, I just trimmed those as aren't part of this patch)

**Comment by Demonid** [comment]
@ariel-  Will this commit fix or enable a way to fix this? https://github.com/TrinityCore/TrinityCore/issues/10840

**Comment by Rushor** [comment]
it's used there https://github.com/TrinityCore/TrinityCore/pull/14454/files#diff-9b5833ba9a7ebbbc5cda78b8a5bbb118L1469, so probably yes @Demonid

**Comment by Frytiks** [comment]
Ready for merge ?

**Comment by Nyeriah** [comment]
Is this ready? Anything left to be done here or is it safe to merge?

**Comment by ariel-** [comment]
It is for me, but let's just wait for the big guys.

**Comment by ccrs** [comment]
@ariel- AKA: Shauren XD

**Comment by tkrokli** [comment]
How come DDuarte (Nayd) labeled this PR [Branch-3.3.5a] when the title says (6.x)? Is it for both?

**Comment by Aokromes** [comment]
Because it's one port of one 6.x made to 3.3.5a branch.

**Comment by tkrokli** [comment]
Yes. I should L2R, L2P and L2C. Feel free to delete my comments if it feels like spam.

**Comment by ccrs** [comment]
8 days have passed, just commenting to keep this alive :3

**Pull request merged**